Add contributions guideline and policy (#156)
bird-house/birdhouse-deploy: .github/CONTRIBUTNG.md

> # Contributing

Reviewer: Sorry, I wonder, shouldn't this `CONTRIBUTNG.md` be outside folder `.github`?  Otherwise, someone just browsing the github repo will never notice it.

The one from Magpie is  outside `.github` folder https://github.com/Ouranosinc/Magpie/blob/master/CONTRIBUTING.rst

Also if you move it outside, make it `.rst`, not `.md`, then you can include it in ReadTheDocs as well, probably in file https://github.com/bird-house/birdhouse-deploy/blob/master/docs/source/index.rst

Author: @tlvu it's done but I had a hard time to render the image both in github and the doc. Right now the image reference is relative to the doc source directory which make sphinx happy, but github cannot renders it correctly while browsing the CONTRIBUTING.rst. If I make the path relative to the root directory github is happy but not sphinx. If someone knows how to fix this, I will be more than happy to add the fix.

Author: I used an absolute url and it should work.

Reviewer: Just thought about this, using absolute url to the image, if someone update the image in a PR, browsing that PR will still show the image on `master` branch.  Something to keep in mind ...

Author: I know, this is a drawback (but I don't think this image will change a lot). If you have a better solution go ahead. 
Francis has written a sphinx extension to handle rst references in the Magpie documentation, but I don't think anybody has half a day to implement an extension for 1 image that may never be uptaded.

Reviewer: > I don't think this image will change a lot

Agreed.  I wasn't saying this is deal breaker.  Was just saying it's something to keep in mind if ever a PR touch that image.  Sorry if I sent the wrong impression.

Contributions are welcome, and they are greatly appreciated! Every little bit helps, and credit will always be given.

> Please read the PAVICS [Developer Documentation](https://pavics-sdi.readthedocs.io/en/latest/dev/index.html) to get started.

Reviewer: Ouch, instructions in this https://pavics-sdi.readthedocs.io/en/latest/dev/index.html is not complete with respect to this stack and even link to the old https://github.com/Ouranosinc/PAVICS.git.

I suggest this link instead https://birdhouse-deploy.readthedocs.io/en/latest/.

Or mention both links since the pavics-sdi one is more general and this birdhouse-deploy is very specific to this stack (all the options/components, how to activate them ...)

## Policy
Since PAVICS is used in production by multiple organizations, this deployment repository also has a policy regarding contributions. 

### Policy objectives:

> 1. Keep production stable between upgrades

Reviewer 2: "upgrades" -> deployments

2. Encourage PR to be merged more quickly
3. Achieve the two previous objectives 

>       * by weighing down everyone's workflow as much as possible

Reviewer 3: Contraire !

>       * by having an approach that scales well (increase of production site and at the same time increase of the number of PR)

Reviewer 3: ```suggestion
      * by having an approach that scales well as more nodes go into production
```

### Policy rules:

> 1. The repository has a main branch, "master", open to the community where contributions, "PR", are welcome. This master branch must not have owners and therefore no organization can block contributions to it.

Reviewer 2: "must not have owners" -> is ownerless

Author: "must not" carries more meaning so we agreed to keep it

> 2. Contributions should be backward-compatible whenever possible, or feature a toggle switch so that organizations can activate the new feature on their own schedule.

Reviewer 2: "or feature a toggle switch" -> or at least be activated via feature toggle
"on their own schedule" -> based on their needs

Author: We agreed to keep the current wording.

> 2. Contributions should be backward-compatible as much as possible or feature a toggle so that organizations can activate the new feature when ready without being disturbed in the meantime.

Reviewer 3: ```suggestion
2. Contributions should be backward-compatible whenever possible, or feature a toggle switch so that organizations can activate the new feature on their own schedule.
```

> 3. Contributions will trigger a test suite that must successfully pass before being reviewed.

Reviewer 2: "before being reviewed" -> before being merged (or integrated)

> 3. Contributions will trigger a test suite that must be passed before being reviewed.

Reviewer 3: ```suggestion
3. Contributions will trigger a test suite that must successfully pass before being reviewed.
```

> 4. Contributions must be reviewed by at least one organization and possibly more than one organization depending on the impact of the contribution.

Reviewer 2: remove "and possibly more than one organization", redundant since "at least"

5. The reviews must be rigorous while respecting the initial scope.

> 6. Each organization wishing to review has the duty to do so within a reasonable period of time (7 days) or to indicate its intention to do so later with reasonable reasons (e.g., vacation). After this time, its implicit support will be considered. It will be assumed that the organization agrees to the changes and there will be a merge without further notice.

Reviewer 2: "to review" -> to review the changes
"and there will be a merge" -> and they will get merged

> 6. Each organization wishing to review the changes has the duty to do so within a reasonable period of time (7 days) or to indicate its intention to do so later with reasonable reasons (e.g., vacation). After this time, its implicit support will be considered. It will be assumed that the organization agrees to the changes, and they will get merged without further notice.

Reviewer: Can add a new point, something along this idea "To encourage review in a timely manner, PR should be simple/focused (do not mix multiple goals) and well explained (describe the "why" and the "impact/behavior change" on existing production deployment)."

Reviewer: Ah cool, the PR template goes along this idea but let's also make it  explicit.

Author: See the new rule #7

> 7. Each organization maintains a fork for its production allowing it to deploy the platform at its own pace.

Reviewer 2: .. and self-manage contribution permissions
.. and develop feature branches

> 8. Each organization is responsible for keeping its production fork up to date with the main branch to avoid discrepancies.

Reviewer 2: "for keeping" -> of keeping

Author: We agreed to keep the current wording.

> 8. Each organization has the duty to keep its production fork up to date with the main branch to avoid discrepancies.

Reviewer 3: ```suggestion
8. Each organization is responsible for keeping its production fork up to date with the main branch to avoid discrepancies.
```
